### PR TITLE
newdevice: Added support for Dasan NOS #5179 + disco change

### PR DESCRIPTION
**Pull request by laf**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

Fixes: #5179 

Also adds support for discovery devices via sysObjectId and sysDescr.

@librenms/reviewers

### CODE REVIEW

File: includes/functions.php [hunk 1]
```diff
@@ -107,6 +107,22 @@ function getHostOS($device)
         }
     }
 
+    $pattern = $config['install_dir'] . '/includes/definitions/*.yaml';
+    foreach (glob($pattern) as $file) {
+        $tmp = Symfony\Component\Yaml\Yaml::parse(
+            file_get_contents($file)
+        );
+        if ($tmp['discovery']['method'] === 'sysObjectId' && is_array($tmp['discovery']['sysObjectId'])) {
```
Comment (murrant) [inline includes/functions.php]: You don't need 'method', just check if sysObjectId or syDescr isset.

File: includes/functions.php [hunk 2]
```diff
@@ -107,6 +107,22 @@ function getHostOS($device)
...
+            if (starts_with($sysObjectId, $tmp['discovery']['sysObjectId'])) {
+                return $tmp['os'];
+            }
+        } elseif ($tmp['discovery']['method'] === 'sysDescr' && is_array($tmp['discovery']['sysDescr'])) {
+            if (starts_with($sysDescr, $tmp['discovery']['sysDescr'])) {
```
Comment (murrant) [inline includes/functions.php]: starts_with doesn't seem very useful for sysDescr, perhaps str_contains or preg_match

File: includes/functions.php [hunk 3]
```diff
@@ -107,6 +107,22 @@ function getHostOS($device)
...
+    $pattern = $config['install_dir'] . '/includes/definitions/*.yaml';
+    foreach (glob($pattern) as $file) {
+        $tmp = Symfony\Component\Yaml\Yaml::parse(
+            file_get_contents($file)
+        );
+        if ($tmp['discovery']['method'] === 'sysObjectId' && is_array($tmp['discovery']['sysObjectId'])) {
+            if (starts_with($sysObjectId, $tmp['discovery']['sysObjectId'])) {
+                return $tmp['os'];
+            }
+        } elseif ($tmp['discovery']['method'] === 'sysDescr' && is_array($tmp['discovery']['sysDescr'])) {
```
Comment (murrant) [inline includes/functions.php]: Some os require both sysObjectId and sysDescr checks, some allow either or it would be nice to accommodate that somehow.

File: includes/definitions/dasan-nos.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+os: dasan-nos
+text: 'Dasan NOS'
+type: network
+icon: dasan
+over:
+    - { graph: device_bits, text: 'Device Traffic' }
+discovery:
+    method: sysObjectId
+    sysObjectId:
+        - .1.3.6.1.4.1.6296.1.2.5.12
```
Comment (murrant) [inline includes/definitions/dasan-nos.yaml]: Perhaps something like this:
```
discovery:
    - sysObjectId: .1.3.6.1.4.1.6296.
    - sysDescr: 'Description'
    - { sysObjectId: .1.3.6.1.4.1.6296., sysDescr: 'Description' }
``` 

The first two lines would match either sysObjectId or sysDescr.
The third line would require both to match.

File: includes/functions.php (hunk 2 above)
Comment (laf) [inline includes/functions.php]: Can change, will use str_contains.

File: includes/functions.php (hunk 1 above)
Comment (laf) [inline includes/functions.php]: The idea with method would be to expand it out like 'sysDescr_regex' and similar (totally made up).

File: includes/functions.php (hunk 3 above)
Comment (laf) [inline includes/functions.php]: Already know about that but just not expanded the code out, kept it simple at present as a poc.

File: includes/definitions/dasan-nos.yaml (hunk 4 above)
Comment (laf) [inline includes/definitions/dasan-nos.yaml]: How would you do something had an array of sysObjectId to check AND an array of sysDescr?

What about if you want or instead of and? That's the reason I've put a method value in.

Comment (laf) [inline includes/definitions/dasan-nos.yaml]: zynos

quanta

powerconnect

macosx

To be honest, I want to keep the yaml simple along with detection. Maybe it's just worth sticking with matching only on an array of sysDescr or sysObjectId?

Comment (murrant) [inline includes/definitions/dasan-nos.yaml]: My goal would be to keep it simple, but allow some complexity.

**Comment by murrant** [comment]
I threw this together quick without testing and it might be a little hard to follow, but here is basically what I'm thinking for the processing:
```php
if (isset($tmp['discovery']) && is_array($tmp['discovery'])) {
    foreach ($tmp['discovery'] as $item) {
        if (!is_array($item) || empty($item)) {
            break;
        }

        // all items must be true
        $result = true;
        foreach ($item as $key => $value) {
            switch ($key) {
                case 'sysObjectId':
                    $result &= starts_with($sysObjectId, $value);
                    break;
                case 'sysDescr':
                    $result &= str_contains($sysDescr, $value);
                    break;
                case 'sysDescr_regex':
                    $result &= preg_match_each($sysDescr, $value);
                    break;
                default:
            }
        }
        if ($result) {
            return $tmp['os'];
        }
    }
}
```
this means that each array under discovery is tested and all items in that array must match.  The values for each can also be an array, such as:
```
discovery:
    - sysObjectId:
        - .1.3.6.1.4.1.6296.
        - .1.3.6.1.4.1.3443.
```

if this is the case any one of those values can match.  I hope this is understandable, I'm a little sick and might not be expressing myself clearly.

**Comment by laf** [comment]
I've updated to use your code :)

Wasn't sure if by preg_match_each you meant preg_match or preg_match_all so switched to preg_match.

The discovery in this new OS works fine using the code

**Comment by murrant** [comment]
Just a simple function, looked very ugly inline
```php
function preg_match_each($haystack, $needles) {
    foreach((array)$needles as $needle) {
        if (preg_match($needle, $subject) {
            return true;
        }
    }
    return false;
}
}

**Comment by murrant** [comment]
Can we get a unit test for this os??

**Comment by laf** [comment]
Test unit pushed.

**Comment by laf** [comment]
Will add more detection into this so hang fire.

**Comment by laf** [comment]
This is ready to go, can add further support but need a walk from a device so can be added later.

**Comment by murrant** [comment]
@laf Made a quick addition to allow multiple regex statements like the others.  If you can rebase this it is ready to merge I think.

**Comment by laf** [comment]
Rebasing + doing docs.

**Comment by laf** [comment]
Rebased + docs added.
